File: elixir/fql/lib/fql/keyval/class.ex
defmodule Fql.KeyVal.Class do
  @moduledoc """
  Classifies a key-value by the kind of operation it represents.

  ## Classes

  - `:constant` - No variables, used for set or returned by get operations
  - `:vstamp_key` - Constant with VStampFuture in the key (set only)
  - `:vstamp_val` - Constant with VStampFuture in the value (set only)
  - `:clear` - Clear operation
  - `:read_single` - Single key-value read operation
  - `:read_range` - Multiple key-value read operation
  """

  alias Fql.KeyVal

  @type class ::
    :constant
    | :vstamp_key
    | :vstamp_val
    | :clear
    | :read_single
    | :read_range
    | {:invalid, String.t()}

  @doc """
  Classifies the given KeyValue.
  """
  @spec classify(KeyVal.key_value()) :: class()
  def classify(kv) do
    dir_attr = get_attributes_of_dir(kv.key.directory)
    key_attr = merge_attributes(dir_attr, get_attributes_of_tup(kv.key.tuple))
    kv_attr = merge_attributes(key_attr, get_attributes_of_val(kv.value))

    cond do
      # KeyValues should never contain nil
      kv_attr.has_nil ->
        invalid_class(kv_attr)

      # KeyValues should contain at most 1 VStampFuture
      kv_attr.vstamp_futures > 1 ->
        invalid_class(kv_attr)

      # Ensure at most one of: vstamp_futures, has_variable, has_clear
      count_conditions(kv_attr) > 1 ->
        invalid_class(kv_attr)

      # Classification based on attributes
      key_attr.has_variable ->
        :read_range

      kv_attr.has_variable ->
        :read_single

      kv_attr.vstamp_futures > 0 ->
        if key_attr.vstamp_futures > 0 do
          :vstamp_key
        else
          :vstamp_val
        end

      kv_attr.has_clear ->
        :clear

      true ->
        :constant
    end
  end

  # Attributes structure
  defp merge_attributes(attr1, attr2) do
    %{
      vstamp_futures: attr1.vstamp_futures + attr2.vstamp_futures,
      has_variable: attr1.has_variable or attr2.has_variable,
      has_clear: attr1.has_clear or attr2.has_clear,
      has_nil: attr1.has_nil or attr2.has_nil
    }
  end

  defp count_conditions(attr) do
    [
      attr.vstamp_futures > 0,
      attr.has_variable,
      attr.has_clear
    ]
    |> Enum.count(&(&1))
  end

  defp invalid_class(attr) do
    parts =
      []
      |> maybe_add(attr.vstamp_futures > 0, "vstamps:#{attr.vstamp_futures}")
      |> maybe_add(attr.has_variable, "var")
      |> maybe_add(attr.has_clear, "clear")
      |> maybe_add(attr.has_nil, "nil")
      |> Enum.join(",")

    {:invalid, "[#{parts}]"}
  end

  defp maybe_add(list, true, item), do: [item | list]
  defp maybe_add(list, false, _item), do: list

  # Get attributes from directory
  defp get_attributes_of_dir(dir) do
    Enum.reduce(dir, empty_attributes(), fn elem, acc ->
      merge_attributes(acc, get_attributes_of_dir_elem(elem))
    end)
  end

  defp get_attributes_of_dir_elem(%{types: _}), do: %{empty_attributes() | has_variable: true}
  defp get_attributes_of_dir_elem(_), do: empty_attributes()

  # Get attributes from tuple
  defp get_attributes_of_tup(tup) do
    Enum.reduce(tup, empty_attributes(), fn elem, acc ->
      merge_attributes(acc, get_attributes_of_tup_elem(elem))
    end)
  end

  defp get_attributes_of_tup_elem(elem) when is_list(elem) do
    get_attributes_of_tup(elem)
  end

  defp get_attributes_of_tup_elem(:nil) do
    %{empty_attributes() | has_nil: true}
  end

  defp get_attributes_of_tup_elem(%{types: _}) do
    %{empty_attributes() | has_variable: true}
  end

  defp get_attributes_of_tup_elem(:maybe_more) do
    empty_attributes()
  end

  defp get_attributes_of_tup_elem(%{tx_version: _, user_version: _}) do
    # VStamp
    empty_attributes()
  end

  defp get_attributes_of_tup_elem(%{user_version: _}) do
    # VStampFuture
    %{empty_attributes() | vstamp_futures: 1}
  end

  defp get_attributes_of_tup_elem(_), do: empty_attributes()

  # Get attributes from value
  defp get_attributes_of_val(val) when is_list(val) do
    get_attributes_of_tup(val)
  end

  defp get_attributes_of_val(:nil) do
    %{empty_attributes() | has_nil: true}
  end

  defp get_attributes_of_val(%{types: _}) do
    %{empty_attributes() | has_variable: true}
  end

  defp get_attributes_of_val(:clear) do
    %{empty_attributes() | has_clear: true}
  end

  defp get_attributes_of_val(%{tx_version: _, user_version: _}) do
    # VStamp
    empty_attributes()
  end

  defp get_attributes_of_val(%{user_version: _}) do
    # VStampFuture
    %{empty_attributes() | vstamp_futures: 1}
  end

  defp get_attributes_of_val(_), do: empty_attributes()

  defp empty_attributes do
    %{
      vstamp_futures: 0,
      has_variable: false,
      has_clear: false,
      has_nil: false
    }
  end
end

File: rust/keyval/src/class.rs
//! Query classification module.
//!
//! This module classifies a KeyValue by the kind of operation it represents.
//! Classifications include: Constant, VStampKey, VStampVal, Clear, ReadSingle, and ReadRange.

use crate::*;
use std::fmt;

/// Classification of a KeyValue based on its structure and contents.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Class {
    /// KeyValue with no Variable, MaybeMore, Clear, or VStampFuture.
    /// Can be used for set operations or returned by get operations.
    Constant,

    /// Constant KeyValue with a VStampFuture in the key.
    /// Can only be used for set operations.
    VStampKey,

    /// Constant KeyValue with a VStampFuture in the value.
    /// Can only be used for set operations.
    VStampVal,

    /// KeyValue with Clear as its value and no Variable, MaybeMore, or VStampFuture.
    /// Used for clear/delete operations.
    Clear,

    /// KeyValue with Variable as its value but no Variable or MaybeMore in its key.
    /// Returns a single KeyValue.
    ReadSingle,

    /// KeyValue with Variable or MaybeMore in its key.
    /// Returns multiple KeyValues.
    ReadRange,

    /// Invalid KeyValue with conflicting attributes.
    Invalid(String),
}

impl fmt::Display for Class {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Class::Constant => write!(f, "constant"),
            Class::VStampKey => write!(f, "vstampkey"),
            Class::VStampVal => write!(f, "vstampval"),
            Class::Clear => write!(f, "clear"),
            Class::ReadSingle => write!(f, "single"),
            Class::ReadRange => write!(f, "range"),
            Class::Invalid(msg) => write!(f, "invalid[{}]", msg),
        }
    }
}

/// Classify a KeyValue based on its structure.
pub fn classify(kv: &KeyValue) -> Class {
    let dir_attr = get_attributes_of_dir(&kv.key.directory);
    let key_attr = dir_attr.merge(&get_attributes_of_tup(&kv.key.tuple));
    let kv_attr = key_attr.merge(&get_attributes_of_val(&kv.value));

    // KeyValues should never contain nil (in Go sense - we don't have this issue in Rust)
    // But we track it for compatibility
    if kv_attr.has_nil {
        return invalid_class(&kv_attr);
    }

    // KeyValues should contain at most 1 VStampFuture
    if kv_attr.vstamp_futures > 1 {
        return invalid_class(&kv_attr);
    }

    // Ensure at most one of these conditions is true
    let count = [
        kv_attr.vstamp_futures > 0,
        kv_attr.has_variable,
        kv_attr.has_clear,
    ]
    .iter()
    .filter(|&&x| x)
    .count();

    if count > 1 {
        return invalid_class(&kv_attr);
    }

    match () {
        _ if key_attr.has_variable => Class::ReadRange,
        _ if kv_attr.has_variable => Class::ReadSingle,
        _ if kv_attr.vstamp_futures > 0 => {
            if key_attr.vstamp_futures > 0 {
                Class::VStampKey
            } else {
                Class::VStampVal
            }
        }
        _ if kv_attr.has_clear => Class::Clear,
        _ => Class::Constant,
    }
}

/// Attributes describing the characteristics of a KeyValue relevant to classification.
#[derive(Debug, Clone, Copy, Default)]
struct Attributes {
    vstamp_futures: usize,
    has_variable: bool,
    has_clear: bool,
    has_nil: bool,
}

impl Attributes {
    /// Merge attributes of parts to infer attributes of the whole.
    fn merge(&self, other: &Attributes) -> Attributes {
        Attributes {
            vstamp_futures: self.vstamp_futures + other.vstamp_futures,
            has_variable: self.has_variable || other.has_variable,
            has_clear: self.has_clear || other.has_clear,
            has_nil: self.has_nil || other.has_nil,
        }
    }
}

/// Get attributes of a directory.
fn get_attributes_of_dir(dir: &Directory) -> Attributes {
    let mut attr = Attributes::default();
    for element in dir {
        match element {
            DirElement::String(_) => {}
            DirElement::Variable(_) => attr.has_variable = true,
        }
    }
    attr
}

/// Get attributes of a tuple.
fn get_attributes_of_tup(tup: &Tuple) -> Attributes {
    let mut attr = Attributes::default();
    for element in tup {
        let sub_attr = match element {
            TupElement::Tuple(t) => get_attributes_of_tup(t),
            TupElement::Variable(_) => Attributes {
                has_variable: true,
                ..Default::default()
            },
            TupElement::MaybeMore => Attributes {
                has_variable: true,
                ..Default::default()
            },
            TupElement::VStampFuture(_) => Attributes {
                vstamp_futures: 1,
                ..Default::default()
            },
            _ => Attributes::default(),
        };
        attr = attr.merge(&sub_attr);
    }
    attr
}

/// Get attributes of a value.
fn get_attributes_of_val(val: &Value) -> Attributes {
    match val {
        Value::Tuple(t) => get_attributes_of_tup(t),
        Value::Variable(_) => Attributes {
            has_variable: true,
            ..Default::default()
        },
        Value::Clear => Attributes {
            has_clear: true,
            ..Default::default()
        },
        Value::VStampFuture(_) => Attributes {
            vstamp_futures: 1,
            ..Default::default()
        },
        _ => Attributes::default(),
    }
}

/// Create an Invalid class with relevant attributes for debugging.
fn invalid_class(attr: &Attributes) -> Class {
    let mut parts = Vec::new();

    if attr.vstamp_futures > 0 {
        parts.push(format!("vstamps:{}", attr.vstamp_futures));
    }
    if attr.has_variable {
        parts.push("var".to_string());
    }
    if attr.has_clear {
        parts.push("clear".to_string());
    }
    if attr.has_nil {
        parts.push("nil".to_string());
    }

    Class::Invalid(parts.join(","))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_classify_constant() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::Int(42)],
            },
            value: Value::String("test".to_string()),
        };
        assert_eq!(classify(&kv), Class::Constant);
    }

    #[test]
    fn test_classify_clear() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::Int(42)],
            },
            value: Value::Clear,
        };
        assert_eq!(classify(&kv), Class::Clear);
    }

    #[test]
    fn test_classify_read_single() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::Int(42)],
            },
            value: Value::Variable(Variable::any()),
        };
        assert_eq!(classify(&kv), Class::ReadSingle);
    }

    #[test]
    fn test_classify_read_range() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::Int(42), TupElement::MaybeMore],
            },
            value: Value::Variable(Variable::any()),
        };
        assert_eq!(classify(&kv), Class::ReadRange);
    }

    #[test]
    fn test_classify_vstamp_key() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::VStampFuture(VStampFuture { user_version: 0 })],
            },
            value: Value::Int(42),
        };
        assert_eq!(classify(&kv), Class::VStampKey);
    }

    #[test]
    fn test_classify_vstamp_val() {
        let kv = KeyValue {
            key: Key {
                directory: vec![],
                tuple: vec![TupElement::Int(42)],
            },
            value: Value::VStampFuture(VStampFuture { user_version: 0 }),
        };
        assert_eq!(classify(&kv), Class::VStampVal);
    }
}
